lib/bolt_server/transport_app.rb

```diff
@@ -191,20 +192,43 @@ def run_script(target, body)
     end
 
     def in_pe_pal_env(environment)
-      if environment.nil?
-        [400, '`environment` is a required argument']
-      else
-        @pal_mutex.synchronize do
-          pal = BoltServer::PE::PAL.new({}, environment)
-          yield pal
-        rescue Puppet::Environments::EnvironmentNotFound
-          [400, {
-            "class" => 'bolt/unknown-environment',
-            "message" => "Environment #{environment} not found"
-          }.to_json]
-        rescue Bolt::Error => e
-          [400, e.to_json]
+      return [400, '`environment` is a required argument'] if environment.nil?
+      @pal_mutex.synchronize do
+        pal = BoltServer::PE::PAL.new({}, environment)
+        yield pal
+      rescue Puppet::Environments::EnvironmentNotFound
+        [400, {
+          "class" => 'bolt/unknown-environment',
+          "message" => "Environment #{environment} not found"
+        }.to_json]
+      rescue Bolt::Error => e
+        [400, e.to_json]
+      end
+    end
+
+    def in_bolt_project(bolt_project)
+      return [400, '`project_ref` is a required argument'] if bolt_project.nil?
+      project_dir = File.join(@config['projects-dir'], bolt_project)
+      return [400, "`project_ref`: #{project_dir} does not exist"] unless Dir.exist?(project_dir)
+      @pal_mutex.synchronize do
+        project = Bolt::Project.create_project(project_dir)
+        bolt_config = Bolt::Config.from_project(project, {})
+        begin
+          pal = Bolt::PAL.new(bolt_config.modulepath, nil, nil, nil, nil, nil, bolt_config.project)
+        rescue StandardError => e
+          raise e.backtrace.to_s
         end
+        module_path = [BoltServer::PE::PAL::PE_BOLTLIB_PATH, Bolt::PAL::BOLTLIB_PATH, *bolt_config.modulepath]
+        # CODEREVIEW: I *think* this is the only thing we need to make different between bolt's PAL. Is it acceptable
+        # to hack this? Modulepath is currently a readable attribute, could we make it writeable?
```

Comment:
This was one of the primary reasons PE::PAL exists, is there a reason not to update the PE::PAL implementation so that it can take project instead of environment?

Reply:
PE::PAL is very environment focused. For bolt projects we want  both the PAL object as well as the project config in the overall context. Given there is only a single change we actually need it seemed reasonable to just use a *slightly* modified bolt PAL.

Reply:
Yeah, I can understand where you're coming from since we don't need much of the functionality we've written for PE::PAL. 

I'm a bit wary of allowing modulepath as a writable attribute because we don't actually need it to be "writable" in an object sense: we just need to change it the one time, and "writable" attributes to me means they are intended to be changed during the lifetime of the object.

I wonder if there's a way we can change bolt's PAL such that we can tell it not to do the module_path munging and instead take the module_path param to the initializer wholesale.

If we did that, we could potentially upstream it to orch and make our code similar.

Reply:
`Bolt::PAL::initialize` signature is already pretty busy. Will defer to @puppetlabs/bolt on that idea.